### PR TITLE
Restructuring

### PR DESCRIPTION
**Pull request by Nima-Ra**
- [x] Write the standard. @Nima-Ra 
- [x] Write the router canister and make it compatible with the standard. @Nima-Ra @miguelToscano 
- [x] Write the template registry canister and make it compatible with the standard. @Nima-Ra 
- [x] Make the canister registry compatible with the standard. @Nima-Ra 
- [x] Update canister registry's upgrade script. @Nima-Ra 
- [x] Make the token registry compatible with the standard. @Nima-Ra 
- [x] Update token registry's upgrade script. @Nima-Ra 
- [x] Make the NFT registry compatible with the standard. @miguelToscano
- [x] Update NFT registry's upgrade script. @miguelToscano
- [x] Update the router canister and add unit tests. @miguelToscano 
- [x] Make DAB.js compatible with the standard and the new structure. @rocky-fleek
- [x] Document the changes. Warn the community about the deprecated methods. @Nima-Ra @nicolaspoggi
- [x] Test all of the canisters with upgrades and also try calling methods from DFX. @Nima-Ra **[WIP]**
- [x] Ask @qti3e and @jsonsivar for review.

### CODE REVIEW

File: registries/nft/src/nft.rs [hunk 1]
```diff
@@ -24,106 +23,59 @@ fn is_controller(account: &Principal) -> bool {
...
-
-                return Ok(true);
-            }
+    pub fn remove(&mut self, principal_id: &Principal) -> Result<(), OperationError> {
+        if self.0.contains_key(principal_id) {
+            self.0.remove(principal_id);
+            return Ok(());
         }
+
+        Err(OperationError::NonExistentItem)
```
Comment (qti3e) [inline registries/nft/src/nft.rs:74]:
```suggestion
        if self.0.remove(&principal_id).is_some() {
            return Ok(());
        }

        Err(OperationError::NonExistentItem)
```

File: registries/nft/src/nft.rs [hunk 2]
```diff
@@ -139,70 +91,46 @@ fn name() -> String {
...
     if !is_controller(&ic::caller()) {
         return Err(OperationError::NotAuthorized);
-    } else if !validate_url(&canister_info.icon) {
+    } else if !validate_url(&canister_info.thumbnail) {
+        return Err(OperationError::BadParameters);
+    } else if canister_info.frontend.is_some() && !validate_url(&canister_info.frontend.clone().unwrap()) {
+        return Err(OperationError::BadParameters);
+    } else if canister_info.details[0].0 != String::from("standard") {
         return Err(OperationError::BadParameters);
     }
```
Comment (qti3e) [inline registries/nft/src/nft.rs:109]:
```suggestion
    } else if canister_info.details[0].0 != String::from("standard") {
        return Err(OperationError::BadParameters);
    } else if canister_info.details.len() != 1 {
         return Err(OperationError::BadParameters);
    }
```

File: registries/nft/src/upgrade.rs [hunk 3]
```diff
@@ -30,8 +59,16 @@ pub fn pre_upgrade() {
...
-    if let Ok((stable,)) = ic::stable_restore::<(StableStorage,)>() {
-        ic::get_mut::<Registry>().load(stable.db);
+    if let Ok((stable,)) = ic::stable_restore::<(StableStorageV0,)>() {
+        let mut updated_nft_canisters = Vec::with_capacity(stable.db.len());
+
+        for (_key, nft_canister) in stable.db.into_iter().enumerate() {
+            let nft_canister_metadata: NftCanister = nft_canister.1.into();
+
+            updated_nft_canisters.push((nft_canister_metadata.principal_id, nft_canister_metadata));
+        }
```
Comment (qti3e) [inline registries/nft/src/upgrade.rs:69]:
```suggestion
        let updated_nft_canisters = stable.db.into_iter().map(|(_, e)| e.into()).collect();
```

File: registries/tokens/src/tokens.rs [hunk 4]
```diff
@@ -168,60 +88,59 @@ fn name() -> String {
...
+fn add(token: Token) -> Result<(), OperationError> {
+    // Check authorization
     if !is_controller(&ic::caller()) {
         return Err(OperationError::NotAuthorized);
     }
 
-    if !validate_url(&token_info.logo) || !validate_url(&token_info.website) {
+    // Check URLs
+    if !validate_url(&token.thumbnail)
+        || (token.frontend.is_some() && !validate_url(&token.frontend.clone().unwrap()))
```
Comment (qti3e) [inline registries/tokens/src/tokens.rs:105]:
```suggestion
        || !token.frontend.map(validate_url).unwrap_or(true)
```

File: registries/tokens/src/tokens.rs [hunk 5]
```diff
@@ -168,60 +88,59 @@ fn name() -> String {
...
-    if !is_controller(&ic::caller()) {
-        return Err(OperationError::NotAuthorized);
+    // Check details
+    if token.details.len() != 4
+        || token.details[0].0 != String::from("symbol")
+        || token.details[1].0 != String::from("standard")
+        || token.details[2].0 != String::from("total_supply")
+        || token.details[3].0 != String::from("verified")
+        || (token.details[3].1 != String::from("true")
+            && token.details[3].1 != String::from("false"))
```
Comment (qti3e) [inline registries/tokens/src/tokens.rs:123]: DetailValue has `True` and `False` variants

File: registries/tokens/src/upgrade.rs [hunk 6]
```diff
@@ -1,9 +1,58 @@
...
+            description: cs.description,
+            frontend: Some(cs.website),
+            thumbnail: cs.logo,
+            principal_id: Principal::from_text("aaaa").unwrap(),
+            details: vec![
+                (String::from("symbol"), cs.symbol),
+                (String::from("standard"), cs.standard),
+                (
+                    String::from("total_supply"),
+                    cs.total_supply.unwrap_or_else(|| 0).to_string(),
```
Comment (qti3e) [inline registries/tokens/src/upgrade.rs:42]: use DetailValue::U64

File: router/src/router.rs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    ) -> Result<(), OperationError> {
+        self.0.insert(registry_info.principal_id, registry_info);
+        Ok(())
+    }
+
+    pub fn remove(&mut self, principal_id: &Principal) -> Result<(), OperationError> {
+        if self.0.contains_key(principal_id) {
+            self.0.remove(principal_id);
+            return Ok(());
+        }
```
Comment (qti3e) [inline router/src/router.rs:35]:
```suggestion
        if self.0.remove(principal_id).is_some() {
            return Ok(());
        }
```

File: template_registry/src/registry.rs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+// The add method will add new entries to the HashMap
+// This method updates the entry if it already exists
+#[update]
+fn add(canister: Principal, metadata: CanisterMetadata) -> Result<(), Error> {
+    if is_admin(&ic::caller()) {
+        // The caller is one of the admins.
+        // The next step is verifying URLs
+        if metadata.details.len() != 0
+            || !validate_url(metadata.thumbnail.clone())
+            || (metadata.frontend.is_some() && !validate_url(metadata.frontend.clone().unwrap()))
```
Comment (qti3e) [inline template_registry/src/registry.rs:71]:
```suggestion
            || !metadata.frontend.map(validate_url).unwrap_or(true)
```

**Review by qti3e — CHANGES_REQUESTED** [review]
i find you guilty of being a NERD xD

File: registries/nft/src/upgrade.rs (hunk 3 above)
Comment (Nima-Ra) [reply registries/nft/src/upgrade.rs:69]: With this code, I have to define the reverse convertion too. (from v1 to v0).

File: registries/canister_registry/src/upgrade.rs [hunk 9]
```diff
@@ -33,8 +79,20 @@ pub fn pre_upgrade() {
 
 #[post_upgrade]
 pub fn post_upgrade() {
-    if let Ok((stable,)) = ic::stable_restore::<(StableStorage,)>() {
-        ic::get_mut::<CanisterDB>().load(stable.db);
+    if let Ok((stable,)) = ic::stable_restore::<(StableStorageV0,)>() {
+        let mut canister_list = Vec::with_capacity(stable.db.len());
+
+        for (_key, canister_info) in stable.db.into_iter().enumerate() {
```
[the remark this answers is not in the document]
Comment (qti3e) [reply registries/canister_registry/src/upgrade.rs:85]: remove the `)`

File: registries/nft/src/upgrade.rs (hunk 3 above)
Comment (qti3e) [reply registries/nft/src/upgrade.rs:69]: no it won't require that, why?

Comment (Nima-Ra) [reply registries/nft/src/upgrade.rs:69]: well, it does.

File: registries/canister_registry/src/upgrade.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Nima-Ra) [reply registries/canister_registry/src/upgrade.rs:85]: that's not the issue. `canister_info` will contain both the key and the value of the entry, hence, it is not any different from the code that has been already committed.

File: registries/canister_registry/src/registry.rs [hunk 10]
```diff
@@ -133,29 +100,29 @@ fn name() -> String {
 }
 
 #[query]
-fn get_info(canisters: Vec<Principal>) -> Vec<Option<&'static CanisterMetadata>> {
+fn get(canister: Principal) -> Option<&'static CanisterMetadata> {
```
[the remark this answers is not in the document]
Comment (Nima-Ra) [reply registries/canister_registry/src/registry.rs:117]: ?

**Review by qti3e — APPROVED** [review]
LGTM

**Pull request merged**